x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticationService.java

```diff
@@ -690,6 +694,51 @@ void writeAuthToContext(Authentication authentication) {
         private void authenticateToken(AuthenticationToken token) {
             this.consumeToken(token);
         }
+
+        private User maybeConsolidateRolesForUser(User user) {
+            if (User.isInternal(user)) {
+                return user;
+            } else if (isAnonymousUserEnabled && anonymousUser.equals(user) == false) {
+                if (anonymousUser.roles().length == 0) {
+                    throw new IllegalStateException("anonymous is only enabled when the anonymous user has roles");
+                }
+                User userWithMergedRoles = user.withRoles(mergeRoles(user.roles(), anonymousUser.roles()));
+                if (user.isRunAs()) {
+                    final User authUserWithMergedRoles = user.authenticatedUser().withRoles(
+                        mergeRoles(user.authenticatedUser().roles(), anonymousUser.roles()));
+                    userWithMergedRoles = new User(userWithMergedRoles, authUserWithMergedRoles);
+                }
+                return userWithMergedRoles;
+            } else {
+                return dedupUserRoles(user);
+            }
+        }
+
+        private String[] mergeRoles(String[] existingRoles, String[] otherRoles) {
+            Set<String> roles = new LinkedHashSet<>(Arrays.asList(existingRoles));
+            if (otherRoles != null) {
+                Collections.addAll(roles, otherRoles);
+            }
+            return roles.toArray(new String[0]);
+        }
+
+        private User dedupUserRoles(User user) {
+            final Set<String> userRoles = new LinkedHashSet<>(Arrays.asList(user.roles()));
+            User userWithDedupRoles = userRoles.size() == user.roles().length
```

Comment:
Do we have to do the length check ? Can't this always be `user.withRoles(Set.of(user.roles()).toArray(new String[0]));`

Reply:
The length checks are to avoid creation of new User objects when it is not necessary. I think most of the times it is indeed not necessary, i.e. roles are already distinct. I did this mostly for: 1) retain existing behavior; 2) potential optimization.

I added a comment to make this intention clear.